### PR TITLE
[DONT SQUASH] Rework use_texture_alpha

### LINKED ISSUE #10342: Rethink use_texture_alpha changes
(see also https://github.com/minetest/minetest/pull/10122#issuecomment-678461977, this is mostly a reminder to myself)

### Summary
Unlike I previously assumed, `use_texture_alpha` does not just decide between opaque and transparent.

Depending on drawtype, it:
* switches between opaque and transparency (`normal` and since the PR: `mesh`, `nodebox`)
* switches between "basic" alpha and transparency (all others)
* is ignored (unclear if this actually happens, for liquid maybe?)

Where:
name | mt | irrlicht | description
---- | -- | -------- | -----------
opaque | `TILE_MATERIAL_OPAQUE` | `video::EMT_SOLID` | opaque regardless of any texture alpha
basic alpha | `TILE_MATERIAL_BASIC` | `video::EMT_TRANSPARENT_ALPHA_CHANNEL_REF` | opaque if pixel alpha < 128, fully see-through if alpha >= 128 
transparency | `TILE_MATERIAL_ALPHA` | `video::EMT_TRANSPARENT_ALPHA_CHANNEL` | pixel alpha = how much see-through it is

### Possible fix

First, the behaviour should be accurately documented (obvious).

<b>Option 1:</b>
Revert the change, easy but doesn't achieve the initial goal.

<b>Option 2:</b>
Give mods full control if material should be opaque, basic or transparent.

Requires backwards compatibility both on the protocol level (`alpha` in tiledef) and mod level (`use_texture_alpha` in nodedef).
firelike, plantlike, allfaces, glasslike should default to "basic"; normal to "opaque" (like before).
mesh, nodebox should default to "opaque" (this is different) and mods should be required\* to explicitly declare if they want transparency for those.

\* With backwards compatibility period as added in the PR

### PR DESCRIPTION
**Pull request by sfan5**
closes #10342

I legit had to use a spreadsheet so I could implement this correct including all required forward and backward-compatible behavior. Here is it:
table

The "value" column has what is set for `use_texture_alpha` / `alpha` (liquids only) in the nodedef. blend/clip/opaque are the new values introduced by this PR.
The cells in the "old client" and "new client" columns explain what a client does when it encounters a nodedef with the specific value set.
The color coding is as follows: **gray** = no compat required, works; **green** = compat is required and can perfectly recreate the wanted behavior; **orange** = compat is required and *cannot* recreate the exact behaviour, but this is okay; **red** = compat is broken
As already hinted at the "(possible) future client" column shows the behavior that would occur with legacy mods after the compatibility code is removed in a few(?) releases.

## To do

This PR is a Ready for Review.

## How to test

Join a world with minetest_game and find that
* there are 50+ warnings (from client) and ~10 (from server) in the console
* torches, (trap)doors, glass stairs render still correctly

Join a devtest world and find that
* there are no warnings
* various nodes that have to do with alpha also still work

### CODE REVIEW

**Comment by hecktest** [comment]
Please rename `basic, alpha` to `clip, blend`, you'll only confuse people by inventing terminology.

**Comment by sfan5** [comment]
Sure, will do.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -7332,10 +7334,16 @@ Used by `minetest.register_node`.
...
+        use_texture_alpha = ...,
+        -- Specifies how the texture's alpha channel will be used for rendering.
+        -- possible values:
+        -- * "opaque": Node is rendered opaque regardless of alpha channel
+        -- * "clip": A given pixel is either fully see-through or opaque
+        --            depending on the alpha channel being below/above 50% in value
+        -- * "blend": The alpha channel specifies how transparent a given pixel
+        --            of the rendered node is
+        -- The default is "opaque" for drawtypes normal, liquid and flowingliquid;
+        -- "clip" otherwise.
```
Comment (SmallJoker) [inline doc/lua_api.txt:7346]: I think the legacy values should somehow be documented as well, even if it's only possible using general/sloppy wording.

**Review by LoneWolfHT — APPROVED** [review]
Tested, works :+1:

File: doc/lua_api.txt (hunk 1 above)
Comment (SmallJoker) [reply doc/lua_api.txt:7346]: Yes, roughly that. Following might fit into the context: `Boolean value (deprecated): true either sets it to blend or clip, false sets it to clip or opaque mode depending on the draw type.`

File: src/nodedef.cpp [hunk 2]
```diff
@@ -597,10 +608,16 @@ void ContentFeatures::deSerialize(std::istream &is)
...
+
+		u8 tmp = readU8(is);
 		if (is.eof()) /* readU8 doesn't throw exceptions so we have to do this */
 			throw SerializationError("");
-		leveled_max = tmp_leveled_max;
+		leveled_max = tmp;
+
+		tmp = readU8(is);
+		if (is.eof())
+			throw SerializationError("");
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/nodedef.cpp:619]: It's try/catched.

**Review by SmallJoker — COMMENTED** [review]
Tested cases:
:ballot_box_with_check: LOTT game. old server + new client. liquids, plants, glass, leaves OK
:ballot_box_with_check: LOTT game. new server + old client. liquids, plants, glass, leaves OK
:ballot_box_with_check: MTG + moreblocks + stained glass. new server + old client. liquids, plants, glass, leaves OK
:ballot_box_with_check: MTG + moreblocks + stained glass. old server + new client. liquids, plants, glass, leaves OK


Minor difference in water color: old server (left) + new client (right)
grafik

No such difference for new server + old client

**Comment by sfan5** [comment]
That'd be because the alpha value of MTG's water texture is not the same that is written in the nodedef.

**Pull request merged**